adapters/mqtt-vertx/src/main/java/org/eclipse/hono/adapter/mqtt/VertxBasedMqttProtocolAdapter.java

```diff
@@ -183,66 +185,95 @@ private void handleEndpointConnection(final MqttEndpoint endpoint) {
             endpoint.reject(MqttConnectReturnCode.CONNECTION_REFUSED_SERVER_UNAVAILABLE);
 
         } else {
-            endpoint.publishHandler(message -> {
+            endpoint.publishHandler(createMqttPublishMessageHandler(endpoint));
 
-                LOG.trace("received message [client ID: {}, topic: {}, QoS: {}, payload {}]", endpoint.clientIdentifier(), message.topicName(),
-                        message.qosLevel(), message.payload().toString(Charset.defaultCharset()));
+            endpoint.closeHandler(v -> {
+                LOG.debug("connection closed with client [{}]", endpoint.clientIdentifier());
+                if (registrationAssertions.remove(endpoint) != null)
+                    LOG.trace("removed registration assertion for client [{}]", endpoint.clientIdentifier());
+            });
+
+            if (getConfig().isAuthenticateDevices()) {
 
+                // check credentials for valid authentication
+                // so far, only hashed-password supported, more to follow
                 try {
 
-                    final ResourceIdentifier resource = ResourceIdentifier.fromString(message.topicName());
+                    MqttUsernamePassword authObject = MqttUsernamePassword.create(endpoint,
+                            getConfig().isSingleTenant());
 
-                    // if MQTT client doesn't specify device_id then closing connection (MQTT has now way for errors)
-                    if (resource.getResourceId() == null) {
-                        close(endpoint);
-                    } else {
+                    Future<Void> validationTracker = Future.future();
+                    validationTracker.setHandler(result -> {
+                        if (result.failed()) {
+                            endpoint.reject(MqttConnectReturnCode.CONNECTION_REFUSED_NOT_AUTHORIZED);
+                        }
+                    });
+                    validateCredentialsForDevice(authObject.getTenantId(), authObject.getType(), authObject.getAuthId(),
+                            authObject.getPassword()).compose(
+                            deviceId -> {
+                                LOG.trace("successfully authenticated device id <{}>", deviceId);
+                                endpoint.accept(false);
+                            }, validationTracker);
+                } catch (IllegalArgumentException e) {
```

Comment:
Well, what we loose then is the exact error message which I tried to make as exact as possible.
I could use a logger for the MqttUsernamePassword class of course, and log there directly, but this feels ugly as well (since it is only a small helper class).
So for now I would stick to the IllegalArgumentException. WDYT?